fletcher/algorithms/string.py

```diff
@@ -267,6 +268,79 @@ def _text_contains_case_sensitive(data: pa.Array, pat: str) -> pa.Array:
     )
 
 
+def _zfill_nonnull(
+    length: int,
+    data: pa.Array,
+    offsets: np.ndarray,
+    data_buffer: np.ndarray,
+    width: int,
+    str_builder: StringArrayBuilder,
+) -> None:
+    for row_idx in range(length):
+        val = data[row_idx]
+        value = np.concatenate(
+            (
+                np.frombuffer(
+                    # as_py: get the number of chars (instead of the len of string data)
+                    (max(0, width - len(val.as_py())) * b"0"),
+                    dtype=np.uint8,
+                ),
+                data_buffer[offsets[row_idx] : offsets[row_idx + 1]],
+            ),
+            axis=0,
+        )
+        str_builder.append_value(value, len(value))
+
+
+def _zfill_nulls(
+    length: int,
+    data: pa.Array,
+    valid_bits: np.ndarray,
+    valid_offset: int,
+    offsets: np.ndarray,
+    data_buffer: np.ndarray,
+    width: int,
+    str_builder: StringArrayBuilder,
+) -> None:
+    for row_idx in range(length):
+        # Check whether the current entry is null.
+        byte_offset = (row_idx + valid_offset) // 8
+        bit_offset = (row_idx + valid_offset) % 8
+        mask = np.uint8(1 << bit_offset)
+        valid = valid_bits[byte_offset] & mask
+
+        if not valid:
+            str_builder.append_null()
+            continue
+
+        val = data[row_idx]
+        value = np.concatenate(
```

Comment:
@xhochy: the question in the draft refers to this part (and the identical part below): specifically the use of `as_py` and whether you have any general ideas/guidance on a better way of doing this. When we take the length of the raw data in the buffer, we run into issues (i.e. with some special characters whose raw representation is longer than 1 char but which should be counted as only a single character towards the string's length).

Reply:
Don't use `data` but rather extract the `value` from `data_buffer` using the offsets in `offsets`. To enable the usage of `numba`, you shouldn't need to pass in the `data: pa.Array,` argument here.

Reply:
A good trick to count the length of an UTF8-string is to count all bytes where the bit-pattern is not `11xxxxx`.

Reply:
A good trick to count the length of an UTF8-string is to count all bytes where the bit-pattern is not `10xxxxx`.

`11xxxxx` is the start of an UTF-8 character
`00xxxxx` is an ASCII character

Reply:
This should help you to get rid of any usage of `data` and instead use `data_buffer`.

Reply:
thanks! will get started on that and will probably have some follow up questions later 😄